### ISSUE TITLE
Durable Entities suddenly stopped de-queuing messages from Control Queue

### ISSUE THREAD

**Issue by jeromekol** [Azure/azure-functions-durable-extension#1213]
### Description
We use an Azure Function (on an dedicated App Service Plan) to consume messages from IoT Hub (~20 messages / second). On this messages we track some properties which we keep state for. This is done with Durable Functions. Today we had the issue that messages were pushed to the Control Queue (SignalEntityAsync()) but were not de-queued anymore. Messages were piling in the control queues until I restarted the Azure function App. This triggered the function again to dequeue the messages.
We are using App Insights for monitoring, but not a single exception has been logged. I just see in the Azure Portal, that the load was high on the CPU and Memory of the App Service Plan.

### Expected behavior
Consistent dequeuing of the messages or an appropriate Exception.

### Actual behavior
Suddenly stopped dequeuing the messages.

### Relevant source code snippets
Queueing to Control Queues: 
```csharp
var entityId = new EntityId(nameof(StatisticsEntity), device.Uid);
await durableEntityClient.SignalEntityAsync(entityId, "operation", values).ConfigureAwait(false);
```
Dequeuing the Control Queues:
```csharp
[FunctionName(nameof(StatisticsEntity))]
public static void HandleOperation([EntityTrigger] IDurableEntityContext context)
{
            var state = context.GetState(() => new StatisticsInfo());
            var input = context.GetInput<Values>();
            // some more logic...
            context.SetState(state);
}
```

### Known workarounds
Reboot the function app

### App Details

- **Durable Functions extension version**: 2.1.1
- **Azure Functions runtime version**:  2.0
- **Programming language used**: C#

### Screenshots
Metric of the related storage account
![image](https://user-images.githubusercontent.com/35995674/74407652-b463eb80-4e32-11ea-9853-d8f1cbc5fc1c.png)

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**: Feb 13, 03:00 AM UTC+1
- **Azure region**: West Europe
- **Orchestration instance ID(s)**:  c81fcb20f2994fac83428cffde949a11, 9d5d8a4cd0184918817487d3be668377 (there are more in case needed)

> If needed, I can share the function app or storage account name privately.

**Comment by cgillum** [comment]
@jeromekol I wasn't able to find IDs "c81fcb20f2994fac83428cffde949a11" or "9d5d8a4cd0184918817487d3be668377" in our telemetry in West Europe in the last 7 days. Are these the right IDs?

Another thing that might help...would you be willing to tell us the first few characters in the name of your storage account? We can search for partial string matches as well.

**Comment by jeromekol** [comment]
@cgillum please try this storage account that starts with buhleruatwe. I found this ID there: {"InstanceId":"@statisticsentity@23ad4e42-2581-4bad-8a0a-3e026de6b702_de-kleie-scale","ExecutionId":"1e6c34749ca147d1a9528456502b1fc3"} hope that works.

**Comment by cgillum** [comment]
Thanks @jeromekol, we found the data we're looking for and are looking into the issue. I can confirm that your app got into a bad state where it stopped picking up new messages and we're currently trying to reason about how this could have happened.

If you happen to run into it again and can [capture a process dump](http://jsandersblog.azurewebsites.net/2017/02/02/how-to-get-a-full-memory-dump-in-azure-app-services/) and can share it with us privately, that could potentially help us root-cause this more quickly.

**Comment by jeromekol** [comment]
Thank you @cgillum, I've opened a Microsoft Support case in the meantime, as it affected production for us.
Apparently we ran into this issue: https://github.com/Azure/azure-functions-host/issues/2015
Following the proposed workaround (add a fix hostid as appsetting) might have solved our issue. Happy to share the case number with you, in case you are interested.

**Comment by cgillum** [comment]
Yes, if you don't mind sharing the case number, I'd be interested in taking a look at the notes it contained.

**Comment by jeromekol** [comment]
120021922000273, thank you for looking into it and your efforts!

**Comment by ConnorMcMahon** [comment]
This appears to be an issue with our pending orchestration count. Somehow we aren't deprecating the value when we process some messages, resulting in the application thinking that we have hit the maximum number of pending messages we allow, while in reality we have none. The memory dump provided in that incident verifies this.

That is why a restart helps.

**Comment by ConnorMcMahon** [comment]
Tracking this under the root cause issue #1274

**Comment by jriouxrandstadca** [comment]
Hi, the same issue just happened to us this morning even with "Microsoft.Azure.DurableTask.AzureStorage" Version="1.7.5" and "Microsoft.Azure.WebJobs.Extensions.DurableTask" Version="2.2.1". Signal was still working but control queues were full of messages. We had to restart function app for the processing to start again.

**Comment by ConnorMcMahon** [comment]
@jriouxrandstadca 

Could you provide us with an instance id (or entity name/key) and a timestamp in UTC?

**Comment by jriouxrandstadca** [comment]
@ConnorMcMahon 
From a message that is in queue since yesterday, I don't know if this is what you need
`"InstanceId":"@companyaggregateroot@b7319478-99a4-41c1-b543-506c7d5dbddf"`
Insertion time (UTC): Fri, 17 Apr 2020 00:10:58 GMT

We restarted the function app at around 10:25 eastern time this morning and scaled with more instances to process everything. But we still have some control queues that just won't go down

**Comment by ConnorMcMahon** [comment]
@jriouxrandstadca 

I'm taking a look, and from what I can tell that although the issue you are having is identical in symptoms, there is a different root cause. Because of that, I am going to reopen this issue, but leave issue #1274 as closed.

**Comment by ConnorMcMahon** [comment]
@jriouxrandstadca 

It looks like for some reason, after one of the entity batches for that entity "completed", we never actually successfully completed the internal orchestration run that represented that batch. This means those messages never got deleted, and ate into your PendingOrchestrationMessage limit. 

I'm not quite sure how that happened, unless the thread processing it got in a deadlock-esque scenario.

The next time you get into that state, can you [take a memory dump](http://jsandersblog.azurewebsites.net/2017/02/02/how-to-get-a-full-memory-dump-in-azure-app-services/)?

**Comment by jriouxrandstadca** [comment]
@ConnorMcMahon 

Yes I will do that if it happens again. Thank you

**Comment by jublair** [comment]
Hi folks,

I appear to have encountered a similar issue here. 

For one of our singleton orchestrations, the history indicates that the orchestrator had completed its run:

![image](https://user-images.githubusercontent.com/46575433/80749417-3df29380-8adb-11ea-8dc1-973dbea94b46.png)

However, the instances table indicates that the orchestrator is still running:

![image](https://user-images.githubusercontent.com/46575433/80749822-eb65a700-8adb-11ea-9c31-e58fd10a84ca.png)

We're still running on v2.2.0. This issue seems to have arisen during our scheduled load testing, so it's possible that this issue arose due to high memory pressure.

**Comment by YunchuWang** [comment]
Retested against current `dev` (`2aa44a48`, extension 3.15.0) using `Microsoft.Azure.DurableTask.AzureStorage` 2.9.1 and could not reproduce the stall.

The existing batched-signal E2E test passed with extended sessions both enabled and disabled. I also ran a targeted lease-churn stress probe: 6,400 signals across 64 entities, `controlQueueBufferThreshold` 32, one operation per entity batch, all four control partitions loaded, and two observed two-worker lease redistribution/recovery cycles. Every signal sent before and after the lease churn drained successfully.

The confirmed root cause tracked in #1274 was fixed by Azure/durabletask#383 and shipped in Azure Storage provider 1.7.5. The current provider still keeps pending message IDs in service-wide shared state across replacement `ControlQueue` instances and removes them on abandon/delete, preventing the counter drift that stopped dequeuing.

Closing as fixed. The later 2.2.1 report in this thread was identified as a different, unconfirmed failure mode; modern sustained-backlog/OOM behavior is tracked separately in #2469, so any new occurrence should be filed with current version, timestamp, and instance details rather than reopening this issue.
